pyface/data_view/data_models/data_accessors.py

```diff
@@ -0,0 +1,168 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Classes for extracting data from objects
+
+This module provides helper classes for the row table data model and
+related classes for extracting data from an object in a consistent way.
+"""
+
+from abc import abstractmethod
+from collections import Hashable
+
+from traits.api import (
+    ABCHasStrictTraits, Any, Event, Instance, Int, Str, observe
+)
+from traits.trait_base import xgetattr, xsetattr
+
+from pyface.data_view.abstract_data_model import DataViewSetError
+from pyface.data_view.abstract_value_type import AbstractValueType
+from pyface.data_view.value_types.api import TextValue
+
+
+class AbstractDataAccessor(ABCHasStrictTraits):
+    """ Accessor that gets and sets data on an object.
+    """
+
+    #: A human-readable label for the accessor.
+    title = Str()
+
+    #: The value type of the title of this accessor, suitable for use in a
+    #: header.
+    title_type = Instance(AbstractValueType, factory=TextValue)
+
+    #: The value type of the data accessed.
+    value_type = Instance(AbstractValueType)
+
+    #: An event fired when accessor is updated update.  The payload is
+    #: a tuple of the accessor info and whether the title or value changed
+    #: (or both).
+    updated = Event
```

Comment:
I thought about this somewhat, and did specifically consider what you are siggesting.  There are two reasons for the current design:
- the data accessor is inspired by the `RowInfo` class in the column example, and when the column model is implemented not as a library, I'd like it to use data accessors as well (although there'll probably be some need for modification to handle hierarchical accessors): https://github.com/enthought/pyface/blob/master/examples/data_view/column_data_model.py#L25
- the potential for hierarchical accessors in the future is why I am returning the object; that was needed to properly handle updates in the column model. That could be dropped for now, but I think I'd still want to return a tuple.
- What I don't want is a proliferation of event traits, just generally as a design, I'd prefer a single "updated" event with a payload.

I expect that the data accessor class will change a bit as more and more models are implemented and we converge on what the "right" abstraction is for declaring how to get things from underlying data objects.

Reply:
> the data accessor is inspired by the RowInfo class in the column example, and when the column model is implemented not as a library, I'd like it to use data accessors as well

Because it was written as an example that might need to be updated later, I did not review it at the same level as I would for library code :)  The example is preliminary, we should not need to veto a design based on that.

> the potential for hierarchical accessors in the future is why I am returning the object

The object is available in the `event.object` attribute, so we still have access to it.

> What I don't want is a proliferation of event traits, just generally as a design, I'd prefer a single "updated" event with a payload.

Actually..., why not? :) If we really need a catch-all, general "updated" event, we can add one that observes the more specific events. The use case here seems to suggest that we need the more specific events, rather than a general one and then differentiate them again.

On the other hand, this event firing logic is hidden from the public API (I think?). So I think we can do with deferring this. There are currently only two places doing this `if event.new[1] == 'value'` check. If more of this pattern occurs, then we can consider splitting the event up.